src/Microsoft.ML.TimeSeries/SrCnnEntireTransformer.cs

```diff
@@ -0,0 +1,903 @@
+            {
+                var meta = new DataViewSchema.Annotations.Builder();
+                meta.AddSlotNames(_parent.OutputLength, GetSlotNames);
+                var info = new DataViewSchema.DetachedColumn[1];
+                info[0] = new DataViewSchema.DetachedColumn(_parent.OutputColumnName, new VectorDataViewType(NumberDataViewType.Double, _parent.OutputLength), meta.ToAnnotations());
+                return info;
+            }
+
+            public void GetSlotNames(ref VBuffer<ReadOnlyMemory<char>> dst) => _slotNames.CopyTo(ref dst, 0, _parent.OutputLength);
+
+            protected override Delegate MakeGetter(DataViewRow input, int iinfo, Func<int, bool> activeOutput, out Action disposer)
+            {
+                Contracts.AssertValue(input);
+                Contracts.Assert(0 <= iinfo && iinfo < _parent.ColumnPairs.Length);
+
+                var src = default(SrCnnTsPoint);
+                var getSrc = input.GetGetter<SrCnnTsPoint>(input.Schema[ColMapNewToOld[iinfo]]);
+
+                disposer = null;
+
+                ValueGetter<VBuffer<double>> del =
+                    (ref VBuffer<double> dst) =>
+                    {
+                        getSrc(ref src);
+                        if (src == null)
+                            return;
+                        var result = VBufferEditor.Create(ref dst, _parent.OutputLength);
+                        result.Values.Fill(Double.NaN);
+                        Double[] values = new Double[_parent.OutputLength];
+                        if (_parent.AnomalyModel.Inference(src, out values))
+                        {
+                            for (int i = 0; i < result.Values.Length; ++i)
+                            {
+                                result.Values[i] = values[i];
+                            }
+                        }
+                        dst = result.Commit();
+                    };
+
+                return del;
+            }
+        }
+
+        internal class SrCnnEntireModeler : ICanSaveModel
+        {
+            internal const string LoaderSignature = "SrCnnEntireModeler";
+            private IHost _host;
+            private SrCnnEntireTransformer _parent;
+            private Dictionary<long, Double[]> _anomalyDict;
+            private static readonly int _minTrainingPoint = 12;
+            private static readonly int _lookaheadWindowSize = 5;
+            private static readonly int _backAddWindowSize = 5;
+            private static readonly int _avergingWindowSize = 3;
+            private static readonly int _judgementWindowSize = 21;
+            private static readonly double _eps = 1e-8;
+            private static readonly double _deanomalyThreshold = 0.35;
+
+            private static readonly Double[] _factors = new Double[]{
+                    184331.62871148242, 141902.71648305038, 109324.12672037778, 84289.9974713784, 65038.57829581667, 50222.84038287002,
```

Comment:
Can you please add some comments on how these numbers came about? What do they mean? What would need to be done if there was a bug and if this table needs to be updated....etc?

Reply:
> Can you please add some comments on how these numbers came about? What do they mean? What would need to be done if there was a bug and if this table needs to be updated....etc?

Added.

Reply:
Is there a formula for this? If so, can you please add the formula with a description of the variables in the formula?

---
In reply to: [418089567](https://github.com/dotnet/machinelearning/pull/4998#discussion_r418089567) [](ancestors = 418089567)

Reply:
> Is there a formula for this? If so, can you please add the formula with a description of the variables in the formula?
> 
> In reply to: [418089567](https://github.com/dotnet/machinelearning/pull/4998#discussion_r418089567) [](ancestors = 418089567)

Formula added.